File: src/ast-search.js
// Mercury Code - AST (Abstract Syntax Tree) Search
// Provides structural code search using tree-sitter-like regex patterns
// and language-aware symbol extraction without external dependencies.
//
// Since we have zero npm dependencies, this module implements lightweight
// AST-like search using regex patterns that understand common language constructs:
//   - Function/method definitions
//   - Class definitions
//   - Import/export statements
//   - Variable declarations
//   - Type definitions
//   - Decorators/annotations
//
// This is NOT a full AST parser — it uses battle-tested regex patterns
// to extract structure from source code, similar to ctags/universal-ctags.

import { readFile, readdir, realpath } from "node:fs/promises";
import path from "node:path";

// ── Language patterns ─────────────────────────────────────────────────────────

const LANGUAGE_PATTERNS = {
  javascript: {
    extensions: [".js", ".jsx", ".mjs", ".cjs"],
    patterns: {
      function: [
        // function name(...)
        /^[ \t]*(?:export\s+)?(?:async\s+)?function\s+(\w+)\s*\(/gm,
        // const name = (async)? (...) =>
        /^[ \t]*(?:export\s+)?(?:const|let|var)\s+(\w+)\s*=\s*(?:async\s+)?(?:\([^)]*\)|[\w]+)\s*=>/gm,
        // const name = function(...)
        /^[ \t]*(?:export\s+)?(?:const|let|var)\s+(\w+)\s*=\s*(?:async\s+)?function/gm,
      ],
      class: [
        /^[ \t]*(?:export\s+)?class\s+(\w+)/gm,
      ],
      method: [
        // methodName(...) { inside class
        /^[ \t]+(?:static\s+)?(?:async\s+)?(?:get\s+|set\s+)?(\w+)\s*\([^)]*\)\s*\{/gm,
      ],
      import: [
        /^[ \t]*import\s+(?:\{[^}]*\}|\*\s+as\s+\w+|\w+)\s+from\s+['"]([^'"]+)['"]/gm,
        /^[ \t]*(?:const|let|var)\s+\{?[^}]*\}?\s*=\s*require\(['"]([^'"]+)['"]\)/gm,
      ],
      export: [
        /^[ \t]*export\s+(?:default\s+)?(?:class|function|const|let|var|async)\s+(\w+)/gm,
        /^[ \t]*export\s+\{([^}]+)\}/gm,
      ],
      variable: [
        /^[ \t]*(?:export\s+)?(?:const|let|var)\s+(\w+)\s*(?::\s*\w+)?\s*=/gm,
      ],
    },
  },

  typescript: {
    extensions: [".ts", ".tsx"],
    patterns: {
      function: [
        /^[ \t]*(?:export\s+)?(?:async\s+)?function\s+(\w+)\s*[<(]/gm,
        /^[ \t]*(?:export\s+)?(?:const|let|var)\s+(\w+)\s*(?::\s*[^=]+)?\s*=\s*(?:async\s+)?(?:\([^)]*\)|[\w]+)\s*=>/gm,
      ],
      class: [
        /^[ \t]*(?:export\s+)?(?:abstract\s+)?class\s+(\w+)/gm,
      ],
      interface: [
        /^[ \t]*(?:export\s+)?interface\s+(\w+)/gm,
      ],
      type: [
        /^[ \t]*(?:export\s+)?type\s+(\w+)\s*[<=]/gm,
      ],
      enum: [
        /^[ \t]*(?:export\s+)?(?:const\s+)?enum\s+(\w+)/gm,
      ],
      method: [
        /^[ \t]+(?:public\s+|private\s+|protected\s+)?(?:static\s+)?(?:readonly\s+)?(?:async\s+)?(?:get\s+|set\s+)?(\w+)\s*[<(]/gm,
      ],
      import: [
        /^[ \t]*import\s+(?:type\s+)?(?:\{[^}]*\}|\*\s+as\s+\w+|\w+)\s+from\s+['"]([^'"]+)['"]/gm,
      ],
      export: [
        /^[ \t]*export\s+(?:default\s+)?(?:class|function|const|let|var|type|interface|enum|async|abstract)\s+(\w+)/gm,
      ],
      variable: [
        /^[ \t]*(?:export\s+)?(?:const|let|var)\s+(\w+)\s*(?::\s*[^=]+)?\s*=/gm,
      ],
      decorator: [
        /^[ \t]*@(\w+)/gm,
      ],
    },
  },

  python: {
    extensions: [".py"],
    patterns: {
      function: [
        /^[ \t]*(?:async\s+)?def\s+(\w+)\s*\(/gm,
      ],
      class: [
        /^[ \t]*class\s+(\w+)/gm,
      ],
      method: [
        /^[ \t]+(?:async\s+)?def\s+(\w+)\s*\(self/gm,
      ],
      import: [
        /^[ \t]*(?:from\s+(\S+)\s+)?import\s+(.+)/gm,
      ],
      variable: [
        /^(\w+)\s*(?::\s*\w+)?\s*=/gm,
      ],
      decorator: [
        /^[ \t]*@(\w+)/gm,
      ],
    },
  },

  go: {
    extensions: [".go"],
    patterns: {
      function: [
        /^func\s+(\w+)\s*\(/gm,
      ],
      method: [
        /^func\s+\(\w+\s+\*?\w+\)\s+(\w+)\s*\(/gm,
      ],
      type: [
        /^type\s+(\w+)\s+(?:struct|interface|func)/gm,
      ],
      import: [
        /^\s+"([^"]+)"/gm,
      ],
      variable: [
        /^(?:var|const)\s+(\w+)/gm,
      ],
    },
  },

  rust: {
    extensions: [".rs"],
    patterns: {
      function: [
        /^[ \t]*(?:pub(?:\([\w:]+\))?\s+)?(?:async\s+)?fn\s+(\w+)/gm,
      ],
      struct: [
        /^[ \t]*(?:pub(?:\([\w:]+\))?\s+)?struct\s+(\w+)/gm,
      ],
      enum: [
        /^[ \t]*(?:pub(?:\([\w:]+\))?\s+)?enum\s+(\w+)/gm,
      ],
      trait: [
        /^[ \t]*(?:pub(?:\([\w:]+\))?\s+)?trait\s+(\w+)/gm,
      ],
      impl: [
        /^[ \t]*impl(?:<[^>]*>)?\s+(?:(\w+)\s+for\s+)?(\w+)/gm,
      ],
      import: [
        /^[ \t]*use\s+(.+);/gm,
      ],
      variable: [
        /^[ \t]*(?:pub\s+)?(?:static|const)\s+(\w+)/gm,
      ],
      macro: [
        /^[ \t]*(?:pub\s+)?macro_rules!\s+(\w+)/gm,
      ],
    },
  },
};

// Skip these directories during search
const SKIP_DIRS = new Set([
  "node_modules", ".git", ".svn", ".hg", "__pycache__",
  "dist", ".next", "coverage", ".cache", "target",
  "build", "vendor", ".tox", "venv", ".venv",
]);

// ── AST Search Functions ──────────────────────────────────────────────────────

/**
 * Detect the language of a file based on its extension.
 */
function detectLanguage(filePath) {
  const ext = path.extname(filePath).toLowerCase();
  for (const [lang, config] of Object.entries(LANGUAGE_PATTERNS)) {
    if (config.extensions.includes(ext)) return lang;
  }
  return null;
}

/**
 * Extract all symbols from a file's content.
 * Returns an array of { kind, name, line, file } objects.
 */
export function extractSymbols(content, filePath) {
  const lang = detectLanguage(filePath);
  if (!lang) return [];

  const patterns = LANGUAGE_PATTERNS[lang].patterns;
  const symbols = [];

  for (const [kind, regexes] of Object.entries(patterns)) {
    for (const regex of regexes) {
      // Create a fresh RegExp instance per call to avoid race conditions
      // with the shared global `g` flag regex (lastIndex is per-instance)
      const freshRegex = new RegExp(regex.source, regex.flags);
      let match;
      while ((match = freshRegex.exec(content)) !== null) {
        // Find line number
        const before = content.slice(0, match.index);
        const line = before.split("\n").length;

        // Get the captured name (first non-undefined capture group)
        const name = match.slice(1).find((g) => g !== undefined) || "";

        symbols.push({
          kind,
          name: name.trim(),
          line,
          file: filePath,
          match: match[0].trim().slice(0, 100), // Context (truncated)
        });
      }
    }
  }

  // Sort by line number
  symbols.sort((a, b) => a.line - b.line);
  return symbols;
}

/**
 * Search for symbols matching a query across the workspace.
 * @param {string} workspace - Workspace root directory
 * @param {string} query - Symbol name or regex pattern to search for
 * @param {object} [options]
 * @param {string} [options.kind] - Filter by kind: function, class, method, etc.
 * @param {string} [options.language] - Filter by language: typescript, python, etc.
 * @param {number} [options.maxResults=200] - Maximum results to return
 * @returns {Promise<Array<{kind, name, line, file, match}>>}
 */
export async function searchSymbols(workspace, query, options = {}) {
  const { kind, language, maxResults = 200 } = options;

  let queryRegex;
  try {
    queryRegex = new RegExp(query, "i");
  } catch {
    queryRegex = new RegExp(query.replace(/[.*+?^${}()|[\]\\]/g, "\\$&"), "i");
  }

  const results = [];
  /** @type {Set<string>} Track visited real paths to prevent symlink loops */
  const visited = new Set();

  async function walkDir(dirPath) {
    if (results.length >= maxResults) return;

    // Resolve real path and check for symlink loops
    let realDirPath;
    try {
      realDirPath = await realpath(dirPath);
    } catch {
      return;
    }
    if (visited.has(realDirPath)) return;
    visited.add(realDirPath);

    let entries;
    try {
      entries = await readdir(dirPath, { withFileTypes: true });
    } catch {
      return;
    }

    for (const entry of entries) {
      if (results.length >= maxResults) break;

      if (entry.isDirectory()) {
        if (!SKIP_DIRS.has(entry.name)) {
          await walkDir(path.join(dirPath, entry.name));
        }
      } else if (entry.isFile()) {
        const filePath = path.join(dirPath, entry.name);
        const lang = detectLanguage(filePath);
        if (!lang) continue;
        if (language && lang !== language) continue;

        let content;
        try {
          content = await readFile(filePath, "utf-8");
        } catch {
          continue;
        }

        // Skip binary / very large files
        if (content.includes("\0") || content.length > 500000) continue;

        const symbols = extractSymbols(content, filePath);
        for (const sym of symbols) {
          if (results.length >= maxResults) break;
          if (kind && sym.kind !== kind) continue;
          if (queryRegex.test(sym.name)) {
            results.push(sym);
          }
        }
      }
    }
  }

  await walkDir(workspace);
  return results;
}

/**
 * Get the structural outline of a single file.
 * Returns all symbols organized by kind.
 */
export async function fileOutline(filePath) {
  let content;
  try {
    content = await readFile(filePath, "utf-8");
  } catch (err) {
    return `Error reading file: ${err.message}`;
  }

  const symbols = extractSymbols(content, filePath);
  if (symbols.length === 0) {
    const lang = detectLanguage(filePath);
    return lang ? "No symbols extracted (file may be too small or use unusual syntax)." : "Unsupported file type for AST analysis.";
  }

  // Group by kind
  const grouped = {};
  for (const sym of symbols) {
    if (!grouped[sym.kind]) grouped[sym.kind] = [];
    grouped[sym.kind].push(sym);
  }

  const lines = [`Outline of ${path.basename(filePath)} (${symbols.length} symbols):`];
  for (const [kind, syms] of Object.entries(grouped)) {
    lines.push(`\n  ${kind.toUpperCase()} (${syms.length}):`);
    for (const sym of syms) {
      lines.push(`    ${sym.name} (line ${sym.line})`);
    }
  }

  return lines.join("\n");
}

/**
 * Format search results for display.
 */
export function formatSearchResults(results) {
  if (!results || results.length === 0) return "No matching symbols found.";

  const lines = [`Found ${results.length} symbol(s):\n`];
  for (const r of results) {
    const relFile = r.file;
    lines.push(`  ${r.kind.padEnd(12)} ${r.name.padEnd(30)} ${relFile}:${r.line}`);
  }
  return lines.join("\n");
}

File: rust-port/crates/mercury-tools/src/ast_search.rs
//! AST-based code search for Mercury Code.
//! Uses regex patterns to find function/class/method definitions.

use regex::Regex;
use std::collections::HashMap;
use std::path::Path;
use once_cell::sync::Lazy;

/// A symbol found in source code.
#[derive(Debug, Clone)]
pub struct Symbol {
    pub name: String,
    pub kind: SymbolKind,
    pub line: usize,
    pub file: String,
    pub source: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum SymbolKind {
    Function, Class, Method, Interface, Struct, Enum, Trait, Const, Variable, Type, Module,
}

impl std::fmt::Display for SymbolKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Function => write!(f, "function"),
            Self::Class => write!(f, "class"),
            Self::Method => write!(f, "method"),
            Self::Interface => write!(f, "interface"),
            Self::Struct => write!(f, "struct"),
            Self::Enum => write!(f, "enum"),
            Self::Trait => write!(f, "trait"),
            Self::Const => write!(f, "const"),
            Self::Variable => write!(f, "variable"),
            Self::Type => write!(f, "type"),
            Self::Module => write!(f, "module"),
        }
    }
}

/// Language-specific regex patterns for symbol detection.
struct LangPatterns {
    patterns: Vec<(SymbolKind, Regex)>,
}

static JS_PATTERNS: Lazy<LangPatterns> = Lazy::new(|| LangPatterns {
    patterns: vec![
        (SymbolKind::Function, Regex::new(r"(?:export\s+)?(?:async\s+)?function\s+(\w+)").unwrap()),
        (SymbolKind::Class, Regex::new(r"(?:export\s+)?class\s+(\w+)").unwrap()),
        (SymbolKind::Const, Regex::new(r"(?:export\s+)?const\s+(\w+)\s*=").unwrap()),
        (SymbolKind::Method, Regex::new(r"^\s+(?:async\s+)?(\w+)\s*\(").unwrap()),
    ],
});

static PYTHON_PATTERNS: Lazy<LangPatterns> = Lazy::new(|| LangPatterns {
    patterns: vec![
        (SymbolKind::Function, Regex::new(r"^(?:async\s+)?def\s+(\w+)").unwrap()),
        (SymbolKind::Class, Regex::new(r"^class\s+(\w+)").unwrap()),
    ],
});

static RUST_PATTERNS: Lazy<LangPatterns> = Lazy::new(|| LangPatterns {
    patterns: vec![
        (SymbolKind::Function, Regex::new(r"(?:pub\s+)?(?:async\s+)?fn\s+(\w+)").unwrap()),
        (SymbolKind::Struct, Regex::new(r"(?:pub\s+)?struct\s+(\w+)").unwrap()),
        (SymbolKind::Enum, Regex::new(r"(?:pub\s+)?enum\s+(\w+)").unwrap()),
        (SymbolKind::Trait, Regex::new(r"(?:pub\s+)?trait\s+(\w+)").unwrap()),
        (SymbolKind::Type, Regex::new(r"(?:pub\s+)?type\s+(\w+)").unwrap()),
        (SymbolKind::Module, Regex::new(r"(?:pub\s+)?mod\s+(\w+)").unwrap()),
        (SymbolKind::Const, Regex::new(r"(?:pub\s+)?const\s+(\w+)").unwrap()),
    ],
});

static GO_PATTERNS: Lazy<LangPatterns> = Lazy::new(|| LangPatterns {
    patterns: vec![
        (SymbolKind::Function, Regex::new(r"^func\s+(\w+)").unwrap()),
        (SymbolKind::Method, Regex::new(r"^func\s+\([^)]+\)\s+(\w+)").unwrap()),
        (SymbolKind::Struct, Regex::new(r"^type\s+(\w+)\s+struct").unwrap()),
        (SymbolKind::Interface, Regex::new(r"^type\s+(\w+)\s+interface").unwrap()),
    ],
});

static JAVA_PATTERNS: Lazy<LangPatterns> = Lazy::new(|| LangPatterns {
    patterns: vec![
        (SymbolKind::Class, Regex::new(r"(?:public|private|protected)?\s*(?:abstract\s+)?class\s+(\w+)").unwrap()),
        (SymbolKind::Interface, Regex::new(r"(?:public\s+)?interface\s+(\w+)").unwrap()),
        (SymbolKind::Method, Regex::new(r"(?:public|private|protected)\s+(?:static\s+)?(?:\w+\s+)(\w+)\s*\(").unwrap()),
    ],
});

/// Get language patterns based on file extension.
fn get_patterns(ext: &str) -> Option<&'static LangPatterns> {
    match ext {
        "js" | "jsx" | "ts" | "tsx" | "mjs" | "cjs" => Some(&*JS_PATTERNS),
        "py" | "pyw" => Some(&*PYTHON_PATTERNS),
        "rs" => Some(&*RUST_PATTERNS),
        "go" => Some(&*GO_PATTERNS),
        "java" | "kt" | "scala" => Some(&*JAVA_PATTERNS),
        _ => None,
    }
}

/// Search for symbols by name across files.
pub fn search_symbols(root: &Path, query: &str) -> Vec<Symbol> {
    let query_lower = query.to_lowercase();
    let mut results = Vec::new();

    let walker = walkdir::WalkDir::new(root)
        .follow_links(false)
        .into_iter()
        .filter_map(|e| e.ok())
        .filter(|e| e.file_type().is_file());

    for entry in walker {
        let path = entry.path();
        let path_str = path.to_string_lossy();
        if path_str.contains("/.git/") || path_str.contains("/node_modules/") || path_str.contains("/target/") {
            continue;
        }

        let ext = path.extension().and_then(|e| e.to_str()).unwrap_or("");
        let patterns = match get_patterns(ext) {
            Some(p) => p,
            None => continue,
        };

        if let Ok(content) = std::fs::read_to_string(path) {
            for (line_num, line) in content.lines().enumerate() {
                for (kind, regex) in &patterns.patterns {
                    if let Some(cap) = regex.captures(line) {
                        if let Some(name) = cap.get(1) {
                            let name_str = name.as_str();
                            if name_str.to_lowercase().contains(&query_lower) {
                                results.push(Symbol {
                                    name: name_str.to_string(),
                                    kind: kind.clone(),
                                    line: line_num + 1,
                                    file: path.to_string_lossy().to_string(),
                                    source: Some(line.trim().to_string()),
                                });
                            }
                        }
                    }
                }
            }
        }
    }

    results
}

/// Extract all symbols from a single file (file outline).
pub fn file_outline(file_path: &Path) -> Vec<Symbol> {
    let ext = file_path.extension().and_then(|e| e.to_str()).unwrap_or("");
    let patterns = match get_patterns(ext) {
        Some(p) => p,
        None => return Vec::new(),
    };

    let content = match std::fs::read_to_string(file_path) {
        Ok(c) => c,
        Err(_) => return Vec::new(),
    };

    let mut symbols = Vec::new();
    for (line_num, line) in content.lines().enumerate() {
        for (kind, regex) in &patterns.patterns {
            if let Some(cap) = regex.captures(line) {
                if let Some(name) = cap.get(1) {
                    symbols.push(Symbol {
                        name: name.as_str().to_string(),
                        kind: kind.clone(),
                        line: line_num + 1,
                        file: file_path.to_string_lossy().to_string(),
                        source: Some(line.trim().to_string()),
                    });
                }
            }
        }
    }

    symbols
}

/// Extract symbols with their full source code (including body).
pub fn extract_symbols(file_path: &Path, symbol_name: &str) -> Vec<Symbol> {
    let outline = file_outline(file_path);
    outline.into_iter().filter(|s| s.name == symbol_name).collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::TempDir;

    #[test]
    fn test_js_function_detection() {
        let dir = TempDir::new().unwrap();
        let file = dir.path().join("test.js");
        std::fs::write(&file, "export function hello() {\n  return 1;\n}\n\nconst foo = 42;\n").unwrap();
        let symbols = file_outline(&file);
        assert!(symbols.iter().any(|s| s.name == "hello" && s.kind == SymbolKind::Function));
        assert!(symbols.iter().any(|s| s.name == "foo" && s.kind == SymbolKind::Const));
    }

    #[test]
    fn test_rust_detection() {
        let dir = TempDir::new().unwrap();
        let file = dir.path().join("test.rs");
        std::fs::write(&file, "pub struct Foo {}\npub fn bar() {}\npub enum Baz {}\n").unwrap();
        let symbols = file_outline(&file);
        assert!(symbols.iter().any(|s| s.name == "Foo" && s.kind == SymbolKind::Struct));
        assert!(symbols.iter().any(|s| s.name == "bar" && s.kind == SymbolKind::Function));
        assert!(symbols.iter().any(|s| s.name == "Baz" && s.kind == SymbolKind::Enum));
    }

    #[test]
    fn test_python_detection() {
        let dir = TempDir::new().unwrap();
        let file = dir.path().join("test.py");
        std::fs::write(&file, "def hello():\n    pass\n\nclass MyClass:\n    pass\n").unwrap();
        let symbols = file_outline(&file);
        assert!(symbols.iter().any(|s| s.name == "hello" && s.kind == SymbolKind::Function));
        assert!(symbols.iter().any(|s| s.name == "MyClass" && s.kind == SymbolKind::Class));
    }

    #[test]
    fn test_search_symbols() {
        let dir = TempDir::new().unwrap();
        std::fs::write(dir.path().join("a.js"), "function fooBar() {}\n").unwrap();
        std::fs::write(dir.path().join("b.js"), "function bazQux() {}\n").unwrap();
        let results = search_symbols(dir.path(), "foo");
        assert_eq!(results.len(), 1);
        assert_eq!(results[0].name, "fooBar");
    }

    #[test]
    fn test_extract_symbols() {
        let dir = TempDir::new().unwrap();
        let file = dir.path().join("test.rs");
        std::fs::write(&file, "fn foo() {}\nfn bar() {}\nfn foo_helper() {}\n").unwrap();
        let results = extract_symbols(&file, "foo");
        assert_eq!(results.len(), 1);
    }

    #[test]
    fn test_unknown_extension() {
        let dir = TempDir::new().unwrap();
        let file = dir.path().join("test.xyz");
        std::fs::write(&file, "some content").unwrap();
        let symbols = file_outline(&file);
        assert!(symbols.is_empty());
    }
}
